File: Project_4/BT9/cpp/bfs_multi.cpp
#include <iostream>
#include <vector>
#include <queue>

// Breadth-first search algorithm for a finite multigraph (no loops, repeated edges allowed)
// Author: Nguyễn Ngọc Thạch
// This program performs BFS traversal on a multigraph starting from a given vertex.

using namespace std;

// BFS function for multigraph
void bfs(const vector<vector<int>>& adj, int start) {
    int n = adj.size();
    vector<bool> visited(n, false);
    queue<int> q;
    visited[start] = true;
    q.push(start);
    while (!q.empty()) {
        int vertex = q.front();
        q.pop();
        cout << vertex << " ";
        for (int neighbor : adj[vertex]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
}

int main() {
    int n, m;
    cout << "Enter number of vertices and edges: ";
    cin >> n >> m;
    vector<vector<int>> adj(n);
    cout << "Enter edges (u v) for each edge (0-indexed vertices, no loops):\n";
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        if (u == v) {
            cout << "Loops are not allowed in a multigraph. Skipping edge (" << u << ", " << v << ").\n";
            continue;
        }
        adj[u].push_back(v);
        adj[v].push_back(u); // Undirected multigraph
    }
    int start;
    cout << "Enter starting vertex: ";
    cin >> start;
    cout << "BFS traversal: ";
    bfs(adj, start);
    cout << endl;
    return 0;
}

File: Project_4/BT9/py/bfs_multi.py
from collections import deque

# Breadth-first search algorithm for a finite multigraph (no loops, repeated edges allowed)
# Author: Nguyễn Ngọc Thạch
# This program performs BFS traversal on a multigraph starting from a given vertex.

# BFS function for multigraph
def bfs(adj, start):
    n = len(adj)
    visited = [False] * n
    q = deque()
    visited[start] = True
    q.append(start)
    while q:
        vertex = q.popleft()
        print(vertex, end=" ")
        for neighbor in adj[vertex]:
            if not visited[neighbor]:
                visited[neighbor] = True
                q.append(neighbor)

def main():
    n, m = map(int, input("Enter number of vertices and edges: ").split())
    adj = [[] for _ in range(n)]
    print("Enter edges (u v) for each edge (0-indexed vertices, no loops):")
    for _ in range(m):
        u, v = map(int, input().split())
        if u == v:
            print(f"Loops are not allowed in a multigraph. Skipping edge ({u}, {v}).")
            continue
        adj[u].append(v)
        adj[v].append(u)  # Undirected multigraph

    start = int(input("Enter starting vertex: "))
    print("BFS traversal: ", end="")
    bfs(adj, start)
    print()

if __name__ == "__main__":
    main()
